### DISCUSSION TITLE
Question about security of encrypted pdf files

### DISCUSSION THREAD

**Discussion by milsap1234** [qpdf/qpdf#530] [Q&A] [answered]
Hello,
Many thanks for qpdf and for the detailed explanations about 256 bit key (https://github.com/qpdf/qpdf/issues/351)
I have a naive general question about security of encrypted pdf's that I hope you will be kind enough to answer.  I read [here](https://security.stackexchange.com/questions/95781/what-security-scheme-is-used-by-pdf-password-encryption-and-why-is-it-so-weak) that 40-bit encryption is "a joke" but it is "now possible to make strongly-encrypted PDF files".  Apart from "a lot of the document structure" being visible, is there any reason to think that the result of

`qpdf --encrypt $pw "" 128 --use-aes=y -- doc.pdf encrypted-doc.pdf`

is less secure than the same document encrypted with the same password by another method, say 

`gpg --symmetric --cipher-algo AES128 doc.pdf ` ?  

Thanks for your insight.

**Comment by jberkenbilt** [comment, marked as answer]
I can try to provide some additional insight. First, one thing to note: In the exact example you provided, the PDF file is not secure at all because you have specified a blank owner password, which means the file can be trivially decrypted. So the rest of this will assume that you use a command where one of user/owner password is `$pw` and the other password is some other password that is at least as strong, such that `$pw` ends up being the weaker of the two passwords.

In that case, what I'll do is try to describe a little about how PDF encryption works. I may end repeating some of what I said #351, but I'm not going to go back and reread all that. :-)

There are two main points here: one is what is actually happening in an encrypted PDF vs. using a general purpose tool like gpg to encrypt the whole file, and the other is how the password is used to generate the encryption key.

First, what is encrypted PDF? A PDF file is a document with a structure that consists of streams, arrays, dictionaries, strings, boolean values, numbers, and this thing called a name, which you can think of as a map key. When you encrypt a PDF, only strings and streams are actually encrypted. Each string and stream are encrypted using a separate key that is derived from a combination of the overall key and the object number. As such, if you were somehow able to brute-force a single string or stream, you would only get that string or stream, but if you were able to brute force the overall key, you could decrypt the whole file. If you have a PDF file that uses object streams and has as much of the file as possible in object streams (with qpdf, use `--object-streams=generate`), then a lot more of the file will be encrypted. At that point, I think there's probably not a whole lot of difference in the level of security between having qpdf implement the regular PDF encryption algorithms vs. using something like gpg.

The second point is about how the password is used. Both gpg and the PDF encryption algorithms will use some approach to derive a key from the password. I have no idea what gpg actually does here. With PDF, the algorithm involves all sorts of hashing of various things, including encryption flags, the password, the document ID, and other things. The actual algorithm is completely different when you use 256-bit encryption, so a PDF encrypted with 256-bit encryption is stronger in more ways than just having a longer key. In principal, if you wanted to use brute force to find the encryption key for a PDF file, you could do so by bypassing the whole password to key thing and going straight to the key. qpdf actually lets you do this with the `--password-is-hex-key` argument. This makes it more or less trivial to break encryption a 40-bit file. It would take considerably more computational power to brute force a 128-bit key and, in that case, you may have better luck by brute-forcing the password, though it depends on the strength of the password. I imagine the same can be said about gpg, but I don't actually know the details of that. For example, I know that, with PDF, a weak password doesn't mean a weak key, it just means that you can find the key by guessing passwords. I suspect the same is true of gpg, but I don't actually know that for sure.

Then another issue is what the purpose of encryption is. As I'm pretty sure I explained thoroughly in #351, PDF uses security to enforce certain restrictions on the file. If you use an encrypted file with no password, you may as well not encrypt the file at all, and it's quite trivial to bypass the restrictions. qpdf will happily do it for you. But it requires you to have the password or the key.

Finally, it's worth noting that there are stronger algorithms than 128-bit AES. If you use PDF encryption, you are limited to what options are available there. If you use an external tool like gpg, you have a potentially wider range of encryption options. As to which ones are better, that's out of my level of knowledge. I'd have to go research it.

So, bottom line, I would say that the security of using gpg with 128-bit AES encryption is comparable to using an external program, though the external program probably gives you a slight edge. More important is which algorithm you use. With PDF, you get significantly better security with 256-bit than 128-bit, but you can do better than that with stronger algorithms and an external program. That said, if you use PDF encryption and a strong password, then you don't need anything other than a compliant PDF reader, which certainly offers you a practical advantage.

I hope this helps. Let me know if you have any additional questions. And I should give the disclaimer that I'm not a cryptography expert, and there may be nuances about what the exact gpg command you gave is that I am unaware of.

**Reply by milsap1234** [reply]
Jay, 
Thanks for your detailed reply and for moving my post to the right place.  I had completely missed the point that a blank owner password made the file insecure *gulp*.   I'm rerereading [#351](https://github.com/qpdf/qpdf/issues/351)  to try to understand better. 
Cheers

**Reply by milsap1234** [reply]
Jay,
A quick followup.  Based on your information above, I prefer 256-bit encryption.  On my system, MX-linux, I was unable to open a 256-bit-encrypted pdf with the default viewer, qpdfview.  However, the file can be opened with a browser (vivaldi or firefox). :)   [post on MX forum](https://forum.mxlinux.org/viewtopic.php?f=97&p=639015#p639015)
Thanks again

**Reply by milsap1234** [reply]
Hello again, Jay.
I have one more comment.  qpdf --help says "Either or both of the user password and the owner password may be empty strings." I naively thought that the user password was to open and read, the owner password to modify once opened with the user password.  I was shocked to learn that " a blank owner password... means the file can be trivially decrypted. "!  I suggest that for users at my level, it would be a lifesaver to include that information in the help message.  
I have version 6.0.0, so if you have already done it, thanks!
And thanks again for your very educational reply to my initial question.  I have quoted you in several places already :)

**Reply by jberkenbilt** [reply]
This question and similar questions come up with some regularity, and this is an area of considerable confusion. I think I will write an appendix to the qpdf manual explaining a little about qpdf encryption in general and how qpdf deals with in particular. Thanks for the suggestion. I'll add it to my TODO list
